dm/master/shardddl/pessimist.go

```diff
@@ -67,13 +68,59 @@ func (p *Pessimist) Start(pCtx context.Context, etcdCli *clientv3.Client) error
 	p.mu.Lock()
 	defer p.mu.Unlock()
 
+	rev1, rev2, err := p.buildLocks(etcdCli)
+	if err != nil {
+		return err
+	}
+	ctx, cancel := context.WithCancel(pCtx)
+	p.wg.Add(1)
+	go func() {
+		defer p.wg.Done()
+		// TODO: handle fatal error from run
+		p.run(ctx, etcdCli, rev1, rev2)
+	}()
+
+	p.closed = false // started now.
+	p.cancel = cancel
+	p.cli = etcdCli
+	p.logger.Info("the shard DDL pessimist has started")
+	return nil
+}
+
+func (p *Pessimist) run(ctx context.Context, etcdCli *clientv3.Client, rev1, rev2 int64) error {
+	for {
+		err := p.watchInfoOperation(ctx, etcdCli, rev1, rev2)
+		if etcdutil.IsRetryableError(err) {
+			retryNum := 1
+			succeed := false
+			for !succeed {
+				select {
+				case <-ctx.Done():
+					return nil
+				case <-time.After(500 * time.Millisecond):
+					rev1, rev2, err = p.buildLocks(etcdCli)
+					if err != nil {
+						log.L().Error("resetWorkerEv is failed, will retry later", zap.Error(err), zap.Int("retryNum", retryNum))
+					}
+					succeed = true
```

Comment:
if `err != nil`, do we still set `succeed = true`?

Reply:
revised in https://github.com/pingcap/dm/pull/555/commits/0df853328942e5cc51268c028a24d7a2f2b3c499